### ISSUE TITLE
RFC: Unify what `MyInductiveType.elim` means

### ISSUE THREAD

**Issue by YaelDillies** [leanprover/lean4#5096]
### Proposal

Currently, the `.elim` lemmas are all over the place in terms of argument order:
- `X.elim` with `x : X` as the first argument: `Nonempty.elim` (core), `Or.elim` (core), `Not.elim` (core), `Ne.elim` (core), `Exists.elim` (core), `Option.elim` (core), `Or.elim3` (Mathlib), 
- `X.elim` with `x : X` as the last argument: `And.elim` (core), `Iff.elim` (core), `Sum.elim` (Batteries), `Option.elim'` (Mathlib)

It would be good for the benefit of everyone if these discrepancies were fixed.

I would further like to argue that we should fix these discrepancies by having:
* `X.cases` take `x : X` as the last argument
* `X.casesOn` take `x : X` as the first argument
* giving up on the `elim` name

This would match the naming of `X.rec` and `X.recOn`.

### Community Feedback

This was discussed on Zulip [in Lean 3](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/option.2Eelim/near/278748002) and [in Lean 4](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Option.20mess). Recently, there was #15448 by @urkud.

Here is a summary of what was said:
* Yes, the current situation is indeed a mess
* We definitely want the version with the `X` argument last in cases where we are building functions out of (not-so-large) large combinations of inductive types. Eg `Sum.elim (Sum.elim (fun _ => trivial) fun _ => trivial) fun _ => trivial` is nicer than `fun x => x.elim (fun y => y.elim (fun _ => trivial) fun _ => trivial) fun _ => trivial` when building a term of type `α ⊕ β ⊕ γ → True`. This also shows up in Mathlib vector notation.
* We want the version with the `X` argument first when we want to do a ["poor-man's `match` statement"](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Option.20mess/near/424711769) to be able to see what each branch is after having provided the main premise. But this need can be circumvented by either using
  * dot notation: `myX.elim _ _`
  * a named argument, assuming the main premise is a named argument: `X.elim (x := myX) _ _`

### Impact

Add :+1: to [issues you consider important](https://github.com/leanprover/lean4/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc). If others benefit from the changes in this proposal being added, please ask them to add :+1: to it.

**Comment by YaelDillies** [comment]
I guess that's a question for @digama0. I assume this is because you know the `X` argument but you don't know how many other arguments there are. Also in `elim _ _ _ myX` you can't use `..` notation.

**Comment by nomeata** [comment]
Tangentially, but while working on structural mutual recursion my code would have been easier and more natural if the major argument comes last, and so if we can unify on that, I think that would be good.

It's also the natural order in the sense that the minor arguments (the cases) don't depend on the major argument, and this argument order makes that clearer.

Less tangentially: is there a difference between `.elim` and `.cases`?

**Comment by YaelDillies** [comment]
[Mario claimed that `.elim` was conventionally meant to always be used as the non-dependent version of `.recOn`](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Option.20mess/near/424708545), so it should always be the same as `.casesOn`, not `.cases`. Doesn't really seem like this rule was followed, though...

**Comment by digama0** [comment]
One way that the `X.casesOn` and `X.cases` names alone don't give you what you want is that the autogenerated lemmas with these names give you nonrecursive but still dependent eliminators, while `X.elim` is always nondependent. Are you proposing to have `X.cases(On)` become nondependent?

**Comment by YaelDillies** [comment]
Right, we shouldn't use `.cases[On]` for the nondependent version. I'm happy to use `.elim[On]` (updated the RFC to that), but I originally didn't since `.elimOn` is currently not used anywhere. I guess that doesn't matter.

**Comment by urkud** [comment]
Are there any plans to deal with it? How can I help it go through?

`#xy`: I have no strong opinion on which convention do we use, but I think it would be nice to have a convention and apply it everywhere (at least, core, Batteries, Mathlib). In particular, I want to have only one of `Option.elim` (core) and `Option.elim'` (Mathlib).

**Comment by YaelDillies** [comment]
P-low means the FRO isn't making it happen, but maybe they would be happy to review a PR doing so?
